File: src/ts-packages/sdk/src/session.ts
import type { ApiClient } from "./api-client.js";
import { ContextBuilder } from "./context-builder.js";
import { Conversation } from "./conversation.js";
import type { AgentifiedTool, GetMessagesTool, GetMessagesOptions, GetMessagesResult, PrepareStepFn } from "./types.js";

export class Session {
  private lastPersistedSeq = 0;
  private lastProcessedStepIndex = 0;
  private readonly _discoverTool;
  private _getMessagesTool?: GetMessagesTool;

  readonly conversation: Conversation;

  constructor(
    readonly id: string,
    readonly namespaceId: string,
    /** @internal */ private readonly sdk: ApiClient,
    /** @internal */ private readonly datasetId: string,
    /** @internal */ private readonly registeredTools: AgentifiedTool[],
  ) {
    this.conversation = new Conversation(sdk, datasetId, namespaceId, id);
    this._discoverTool = sdk.asDiscoverTool(datasetId, namespaceId, id);
  }

  get context(): ContextBuilder {
    return new ContextBuilder(
      this.sdk, this.datasetId, this.namespaceId, this.id,
      this.registeredTools,
      this._discoverTool.discoveredNames,
    );
  }

  get discoverTool() {
    return this._discoverTool;
  }

  get getMessagesTool(): GetMessagesTool {
    if (!this._getMessagesTool) {
      this._getMessagesTool = this.sdk.asGetMessagesTool(this.datasetId, this.namespaceId, this.id);
    }
    return this._getMessagesTool;
  }

  async getMessages(opts?: GetMessagesOptions): Promise<GetMessagesResult> {
    const res = await this.sdk.getContext(this.datasetId, this.namespaceId, this.id, {
      strategy: opts?.strategy,
      maxTokens: opts?.maxTokens,
    });
    let messages = res.messages;
    let includedMessages = res.includedMessages;
    if (opts?.maxMessages && messages.length > opts.maxMessages) {
      messages = messages.slice(messages.length - opts.maxMessages);
      includedMessages = messages.length;
    }
    return {
      messages,
      totalMessages: res.totalMessages,
      includedMessages,
      strategyUsed: res.strategyUsed,
      fallback: res.fallback,
    };
  }

  readonly prepareStep: PrepareStepFn = async ({ steps }) => {
    const activeTools = new Set(["agentified_discover", "agentified_get_messages"]);

    const newSteps = steps.slice(this.lastProcessedStepIndex);

    const messages: Array<{ role: string; content: string; tool_calls?: unknown; tool_call_id?: string }> = [];
    for (const step of newSteps) {
      if (step.text) {
        messages.push({ role: "assistant", content: step.text });
      }
      if (step.toolCalls && step.toolCalls.length > 0) {
        messages.push({ role: "assistant", content: "", tool_calls: step.toolCalls });
      }
      if (step.toolResults) {
        for (const result of step.toolResults) {
          const raw = result.result;
          const content = typeof raw === "string" ? raw : JSON.stringify(raw) ?? "";
          messages.push({ role: "tool", content, tool_call_id: result.toolCallId });
        }
      }
    }

    if (messages.length > 0) {
      const res = await this.sdk.appendMessages(this.datasetId, this.namespaceId, this.id, messages);
      this.lastPersistedSeq = res.lastSeq;
    }

    this.lastProcessedStepIndex = steps.length;

    for (const name of this._discoverTool.discoveredNames) activeTools.add(name);

    return { activeTools: [...activeTools] };
  };

  async updateConversation(input: { messages: Array<{ role: string; content: string }> }): Promise<void> {
    const { messages } = input;
    if (messages.length === 0) return;

    const stored = await this.sdk.getMessages(this.datasetId, this.namespaceId, this.id, { limit: messages.length });
    const tail = stored.messages;

    let overlap = 0;
    for (let tryLen = Math.min(tail.length, messages.length); tryLen > 0; tryLen--) {
      const tailStart = tail.length - tryLen;
      let match = true;
      for (let i = 0; i < tryLen; i++) {
        if (tail[tailStart + i]!.role !== messages[i]!.role || tail[tailStart + i]!.content !== messages[i]!.content) {
          match = false;
          break;
        }
      }
      if (match) {
        overlap = tryLen;
        break;
      }
    }

    const newMessages = messages.slice(overlap);
    if (newMessages.length === 0) return;

    const res = await this.sdk.appendMessages(this.datasetId, this.namespaceId, this.id, newMessages);
    this.lastPersistedSeq = res.lastSeq;
  }
}

File: src/py-packages/sdk/src/agentified/session.py
from __future__ import annotations

from typing import Any, TYPE_CHECKING

from .context_builder import ContextBuilder
from .conversation import Conversation
from .models import (
    AgentifiedTool,
    DiscoverTool,
    GetMessagesOptions,
    GetMessagesResult,
    GetMessagesTool,
)

if TYPE_CHECKING:
    from .api_client import ApiClient


class Session:
    def __init__(
        self,
        id: str,
        namespace_id: str,
        sdk: ApiClient,
        dataset_id: str,
        registered_tools: list[AgentifiedTool],
    ) -> None:
        self.id = id
        self.namespace_id = namespace_id
        self._sdk = sdk
        self._dataset_id = dataset_id
        self._registered_tools = registered_tools
        self.conversation = Conversation(sdk, dataset_id, namespace_id, id)
        self._discover_tool = sdk.as_discover_tool(dataset_id, namespace_id, id)
        self._get_messages_tool: GetMessagesTool | None = None

    @property
    def context(self) -> ContextBuilder:
        return ContextBuilder(
            self._sdk, self._dataset_id, self.namespace_id, self.id,
            registered_tools=self._registered_tools,
            discovered_names=self._discover_tool.discovered_names,
        )

    @property
    def discover_tool(self) -> DiscoverTool:
        return self._discover_tool

    @property
    def get_messages_tool(self) -> GetMessagesTool:
        if self._get_messages_tool is None:
            self._get_messages_tool = self._sdk.as_get_messages_tool(
                self._dataset_id, self.namespace_id, self.id,
            )
        return self._get_messages_tool

    async def get_messages(self, opts: GetMessagesOptions | None = None) -> GetMessagesResult:
        strategy = opts.strategy if opts else None
        max_tokens = opts.max_tokens if opts else None
        max_messages = opts.max_messages if opts else None

        res = await self._sdk.get_context(
            self._dataset_id, self.namespace_id, self.id,
            strategy=strategy, max_tokens=max_tokens,
        )
        messages = res.messages
        included = res.included_messages
        if max_messages and len(messages) > max_messages:
            messages = messages[len(messages) - max_messages:]
            included = len(messages)

        return GetMessagesResult(
            messages=messages,
            total_messages=res.total_messages,
            included_messages=included,
            strategy_used=res.strategy_used,
            fallback=res.fallback,
        )

    async def update_conversation(self, messages: list[dict[str, str]]) -> None:
        if not messages:
            return

        stored = await self._sdk.get_messages(
            self._dataset_id, self.namespace_id, self.id, limit=len(messages)
        )
        tail = stored.messages

        overlap = 0
        for try_len in range(min(len(tail), len(messages)), 0, -1):
            tail_start = len(tail) - try_len
            match = True
            for i in range(try_len):
                if tail[tail_start + i].role != messages[i]["role"] or tail[tail_start + i].content != messages[i]["content"]:
                    match = False
                    break
            if match:
                overlap = try_len
                break

        new_messages = messages[overlap:]
        if not new_messages:
            return

        await self._sdk.append_messages(self._dataset_id, self.namespace_id, self.id, new_messages)
